### PR TITLE
Distribute tar file creation across multiple turns of the event loop

### PR DESCRIPTION
**Pull request by humphd**
The code that creates a tarball will sometimes timeout the server, since a big project with lots of (possibly big) files will all need to get processed.  This fix is a first-level attempt to fix that, by splitting the work across multiple turns of the event loop.  By doing so, bits of the work will happen, but not such that it blocks the whole process.  A more elaborate fix would do it in a secondary process, but I want to see if we can improve things with something simpler.

Based on my reading, I think I want `setImmediate` vs. `process.nextTick`, but feel free to tell me I'm wrong if you can back it up!

### CODE REVIEW

**Comment by humphd** [comment]
@jbuck can you r? this for me, too?  Make sure I'm clear on this.

**Comment by gideonthomas** [comment]
I think this looks right!

File: api/classes/base_controller.js [hunk 1]
```diff
@@ -87,27 +87,32 @@ BaseController.prototype.getAllAsMeta = function(req, reply) {
...
         tarStream.entry({ name: file.get('path') }, model.get('buffer'));
       });
-    }).then(function() {
-      tarStream.finalize();
     });
   }
 
-  startStreaming(files, self.Model).catch(errors.generateErrorResponse);
+  setImmediate(function() {
+    Promise.map(files, processFile)
```
Comment (jbuck) [inline api/classes/base_controller.js]: I wonder if you'd also want to specify a [concurrency option](https://github.com/petkaantonov/bluebird/blob/master/API.md#option-concurrency) here, to limit the number of in-flight database queries

**Comment by jbuck** [comment]
@humphd only one possible change here. Let me know what you think of that!

File: api/classes/base_controller.js (hunk 1 above)
Comment (humphd) [inline api/classes/base_controller.js]: I think this sounds like a good idea.  Do you have a suggestion as to what I should set it to?  The docs don't really give much help here, just indicate that you can limit it.  What's a good number, do you think?

Comment (jbuck) [inline api/classes/base_controller.js]: It looks like the default connection pooling setup for bookshelf is 10 maximum connections, so we definitely want it lower than that. Ball-parking it... maybe limit it to 2 simultaneous entries at once?

**Comment by humphd** [comment]
@jbuck, updated.  You cool with this?

**Comment by jbuck** [comment]
@humphd r+ with nits around the style that travis is complaining about

**Comment by humphd** [comment]
Cool, thanks you two.  We should ship this to staging/prod, too.

**Pull request merged**